elastic/kibana
# [Upgrade Assistant] Add support for deprecated index settings

Fixes https://github.com/elastic/kibana/issues/92430

I'm skipping a release note since UA will be disabled on `7.13`.

### How to test
This PR requires some hacking to test on `master` 😅 as we can't test `7.x` deprecations on master (`8.0`). I've added a commit (and reverted it) with the hacks to hopefully ease the testing process. See: https://github.com/elastic/kibana/pull/93293/commits/613bddf2914153d7c77fa626e671a2b54fa0d5aa.

In addition to the code hacks, you'll need to create the indices set in the deprecations response.

```
PUT test1
{
  "settings": {
    "index" : {
      "refresh_interval": "1m"
      }      
  }
}

PUT test2
{
  "settings": {
    "index" : {
      "refresh_interval": "1m"
      }      
  }
}
```

Note: I primarily developed this on the `7.x` branch to ensure everything is working as expected.

There are two parts to this PR:
1. If a user needs to reindex an index *and* the index has deprecated index settings, we should alert them of this and automatically delete them when performing the reindex. (Currently, there are no replacements for any of the deprecated index settings. If this were to change, we could alternatively replace the deprecated setting with the suggested new setting.) To confirm the index setting was removed, you can go to Index Management and check the reindexed index's settings (or use the GET index settings API in Console).

**Related screenshot:**
<img width="399" alt="Screen Shot 2021-03-10 at 9 28 09 AM" src="https://user-images.githubusercontent.com/5226211/110683730-58d96e00-81aa-11eb-9b0c-f6dc09505b6b.png">

2. If a user is using a deprecated index setting *and* it is surfaced through the deprecation info API, then we provide a "Fix" button to remove the setting. This is currently only applicable to [translog retention settings](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-modules-translog.html#index-modules-translog-retention), but the code is written in a way to support any number of index deprecations surfaced in the deprecation info API. To confirm the index setting was removed, you can go to Index Management and check the index settings (or use the GET index settings API in Console).

**Related screenshots:** 
<img width="1250" alt="Screen Shot 2021-03-09 at 3 30 48 PM" src="https://user-images.githubusercontent.com/5226211/110533905-76e09900-80ec-11eb-88eb-f753f008a189.png">
<img width="731" alt="Screen Shot 2021-03-09 at 3 30 39 PM" src="https://user-images.githubusercontent.com/5226211/110533899-76480280-80ec-11eb-85c2-11eb6e220e33.png">
